web/src/main/java/org/springframework/security/web/AroundFilterObservation.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.micrometer.observation.Observation;
+import jakarta.servlet.Filter;
+import jakarta.servlet.FilterChain;
+
+interface AroundFilterObservation extends FilterObservation {
+
+	AroundFilterObservation NOOP = new AroundFilterObservation() {
+	};
+
+	static AroundFilterObservation create(Observation before, Observation after) {
+		if (before.isNoop() || after.isNoop()) {
+			return NOOP;
+		}
+		return new SimpleAroundFilterObservation(before, after);
+	}
+
+	default Observation before() {
+		return Observation.NOOP;
+	}
+
+	default Observation after() {
+		return Observation.NOOP;
+	}
+
+	class SimpleAroundFilterObservation implements AroundFilterObservation {
+
+		private final Iterator<Observation> observations;
+
+		private final Observation before;
+
+		private final Observation after;
+
+		private AtomicReference<Observation> currentObservation = new AtomicReference<>(null);
+
+		private AtomicReference<Observation.Scope> currentScope = new AtomicReference<>(null);
+
+		SimpleAroundFilterObservation(Observation before, Observation after) {
+			this.before = before;
+			this.after = after;
+			this.observations = Arrays.asList(before, after).iterator();
+		}
+
+		@Override
+		public void start() {
+			if (this.observations.hasNext()) {
+				stop();
+				Observation observation = this.observations.next();
+				observation.start();
+				Observation.Scope scope = observation.openScope();
+				this.currentScope = new AtomicReference<>(scope);
```

Comment:
Shouldn't we make the `AtomicReference` `final` and use `currentScope.set(scope)` instead?

Because this is changing the reference of the `currentScope` and visibility is not guaranteed, it's not thread safe. We can either use a `volatile Observation.Scope` or an `AtomicReference` but in that case, the `AtomicReference` should be `final` and `.set(...)` should be used to change the value.